src/hipchatbot.js

```diff
@@ -0,0 +1,198 @@
+import moment from 'moment';
+import App from './models/App';
+import { HIPCHAT_CALLBACK_URL } from './config';
+
+export default class Hipchatter extends HipchatterBase {
+  get_session(callback) {
+    this.request('get', 'oauth/token/'+this.token, callback);
+  }
+}
+
+export default class HipchatBot {
+  constructor(express) {
+    this.express = express;
+    this.start();
+  }
+
+  async start() {
+    this.app = await App.get();
+    if (!this.app.hipchat.oauthId || !this.app.hipchat.oauthSecret) {
+      throw new Error('Before you can run Jeopardy Bot on Hipchat, you must first install it as an add-on.');
+    }
+
+    this.express.get('/capabilities', this.buildCapabilitiesDescriptor.bind(this));
+    this.express.post('/install', this.install.bind(this));
+    await this.validateToken(this.app.hipchat);
+    this.hipchatter = new Hipchatter(this.app.hipchat.accessToken.token);
+    const roomId = await this.getInstalledRoomId();
+    this.registerWebhook(roomId, 'room_message');
+    // this.registerWebhook(roomId, 'room_enter');
+    this.express.post('/webhook', this.onMessage.bind(this));
+  }
+
+  getInstalledRoomId() {
+    return new Promise(resolve => {
+      this.hipchatter.get_session((err, session) => {
+        resolve(session.client.room.id);
+      });
+    });
+  }
+
+  async validateToken(hipchat) {
+    if (!hipchat.accessToken.token || moment().isAfter(hipchat.accessToken.expires_in)) {
+      const { access_token, expires_in } = await this.getAccessToken(hipchat.oauthId, hipchat.oauthSecret);
+      this.app.hipchat.accessToken.token = access_token;
+      this.app.hipchat.accessToken.expires = moment().add(expires_in, 'seconds');
+      await this.app.save();
+    }
+  }
+
+  buildCapabilitiesDescriptor(req, res) {
+    const descriptor = {
+      "name": "Jeopardy Bot",
+      "description": "This is Jeopardy!",
+      "key": "com.jeopardy.bot",
+      "links": {
+        "homepage": "https://github.com/kesne/jeopardy-bot",
+        "self": HIPCHAT_CALLBACK_URL + "/capabilities",
+      },
+      "capabilities": {
+        "hipchatApiConsumer": {
+          "scopes": [
+            "admin_room",
+            "send_notification",
+          ]
+        },
+        "installable": {
+          "callbackUrl": HIPCHAT_CALLBACK_URL + "/install",
+          "allowGlobal": false,
+          "allowRoom": true,
+        },
+      },
+    }
+
+    res.send(descriptor);
+  }
+
+  async install(req, res) {
+    const { capabilitiesUrl, oauthId, roomId, oauthSecret } = req.body;
+    this.app.hipchat.oauthId = oauthId;
+    this.app.hipchat.oauthSecret = oauthSecret;
+
+    let tokenUrl;
+    request(capabilitiesUrl, (err, res, body) => {
+      if (!err && res.statusCode === 200) {
+        tokenUrl = JSON.parse(body).capabilities.oauth2Provider.tokenUrl;
+      } else {
+        this.onError(err);
+      }
+    });
+
+    const { access_token, expires_in } = await this.getAccessToken(oauthId, oauthSecret, tokenUrl);
```

Comment:
I'd rename these as they come in (i.e. `{ access_token: token, expires_in: expires }`). Snake case is so gross.